pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -193,108 +218,151 @@ func (c *EtcdCertSignerController) syncAllMasters(recorder events.Recorder) erro
+	return err
+}
 
-	// create the certificates and update them in the API
-	pCert, pKey, err := tlshelpers.CreatePeerCertKey(etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nodeInternalIPs)
-	if err != nil {
-		return err
+// newCertSecret ensures consistency of secret creation between the controller
+// and its tests.
+func newCertSecret(secretName, nodeUID string, cert, key []byte) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: operatorclient.TargetNamespace,
+			Annotations: map[string]string{
+				nodeUIDAnnotation: nodeUID,
+			},
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"tls.crt": cert,
+			"tls.key": key,
+		},
 	}
+}
 
-	if err := c.createSecret(etcdPeerClientSecretName, pCert, pKey, recorder); err != nil {
-		return err
+// checkCertValidity validates the provided cert bytes. If the cert needs to be
+// regenerated, a message will be returned indicating why. An empty message
+// indicates a valid cert. An error will be returned if the cert is not valid
+// and should not be regenerated.
+func checkCertValidity(certBytes, keyBytes []byte, ipAddresses []string, nodeUIDUnchanged bool) (string, error) {
+	// Loading the keypair without error indicates the key material is valid and
+	// the cert and private key are related.
+	keyPair, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		// Invalid keypair - regen required
+		return fmt.Sprintf("%v", err), nil
 	}
-	sCert, sKey, err := tlshelpers.CreateServerCertKey(etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nodeInternalIPs)
+	leafCert, err := x509.ParseCertificate(keyPair.Certificate[0])
 	if err != nil {
-		return err
+		// Should never happen once x509KeyPair() succeeds
+		return fmt.Sprintf("%v", err), nil
 	}
 
-	if err := c.createSecret(etcdServingSecretName, sCert, sKey, recorder); err != nil {
-		return err
+	// Check that the cert is valid for the provided ip addresses
+	errs := []error{}
+	for _, ipAddress := range ipAddresses {
+		if err := leafCert.VerifyHostname(ipAddress); err != nil {
+			errs = append(errs, err)
+		}
 	}
-
-	metricCert, metricKey, err := tlshelpers.CreateMetricCertKey(etcdMetricCASecret.Data["tls.crt"], etcdMetricCASecret.Data["tls.key"], nodeInternalIPs)
-	if err != nil {
-		return err
+	if len(errs) > 0 {
+		if nodeUIDUnchanged {
+			// This is an error condition. If the cert SAN doesn't include all
+			// of ip's for the node that it was generated for, an out-of-band
+			// etcd cluster membership change may be required. The operator
+			// doesn't currently handle this.
+			return "", utilerrors.NewAggregate(errs)
+		} else {
+			// A different node uid indicates node removal and addition with the
+			// same name. etcd on the node will be a new cluster member and
+			// needs a new certificate.
+			msgs := []string{}
+			for _, err := range errs {
+				msgs = append(msgs, fmt.Sprintf("%v", err))
```

Comment:
what does the error look like here it seems that explicitly calling out the fact that node UID is missing OR changed might be useful for debugging? It would be expected that on an upgrade every cert would be rotated. If this is expected we don't want to scare folks with logs. Curious about your thoughts here.